corehq/motech/fhir/views.py

```diff
@@ -30,3 +42,82 @@ class EditFHIRRepeaterView(EditRepeaterView, AddFHIRRepeaterView):
     @property
     def page_url(self):
         return reverse(self.urlname, args=[self.domain])
+
+
+@require_GET
+@login_or_api_key
+@require_superuser
+@toggles.FHIR_INTEGRATION.required_decorator()
+def get_view(request, domain, fhir_version_name, resource_type, resource_id):
+    fhir_version = _get_fhir_version(fhir_version_name)
+    if not fhir_version:
+        return JsonResponse(status=400, data={'message': "Unsupported FHIR version"})
+    case = get_case_or_404(domain, resource_id)
+
+    if not FHIRResourceType.objects.filter(
+            domain=domain,
+            fhir_version=fhir_version,
+            name=resource_type,
+            case_type__name=case.type
+    ).exists():
+        return JsonResponse(status=400, data={'message': "Invalid Resource Type"})
+    response = {
+        'resourceType': resource_type,
+        'id': resource_id
+    }
+    response.update(build_fhir_resource(case))
+    return JsonResponse(response)
+
+
+@require_GET
+@login_or_api_key
+@require_superuser
+@toggles.FHIR_INTEGRATION.required_decorator()
+def search_view(request, domain, fhir_version_name, resource_type):
+    fhir_version = _get_fhir_version(fhir_version_name)
+    if not fhir_version:
+        return JsonResponse(status=400, data={'message': "Unsupported FHIR version"})
+    patient_case_id = request.GET.get('patient_id')
+    if not patient_case_id:
+        return JsonResponse(status=400, data={'message': "Please pass patient_id"})
+    case_accessor = CaseAccessors(domain)
+    try:
+        patient_case = case_accessor.get_case(patient_case_id)
+        if patient_case.is_deleted:
+            return JsonResponse(status=400, data={'message': f"Patient with ID {patient_case_id} was removed"})
+    except CaseNotFound:
+        return JsonResponse(status=400, data={'message': f"Could not find patient with ID {patient_case_id}"})
+
+    case_types_for_resource_type = list(
+        FHIRResourceType.objects.filter(
+            domain=domain, name=resource_type, fhir_version=fhir_version
+        ).values_list('case_type__name', flat=True)
+    )
+    if not case_types_for_resource_type:
```

Comment:
It might also make this easier to understand if we confirmed that there is only one item in this list ... or maybe used a query that made it obvious, like `get()` instead of `filter()`.

Reply:
umm, so this for the list of other related resources to the main case here and hence we are looking for all types. Does that sound right?

Reply:
I don't think I understand. Can you give me an example?

Reply:
ah, you are right. I misunderstood.

So this is the validation that we should only have one case type per resource type?
Using `get` here sounds good. I will update that. Though we should add that as a validation on the model? or is it a use case that is available via admin end?

Reply:
ah, i just realized this should be `filter` only, since we still do support multiple resource types per case type outside DD.

So what is happening here is,
1. HQ gets a request to find a certain resource type, say `Observation` for a patient with case id `patient_id` sent as a part of the URL
2. HQ first finds the patient case with that ID and if it finds it it moves further
3. HQ then finds the corresponding case types for the resource type being requested
4. It then finds, the cases, of those case types, that are related to the patient case and then return the link to those in a bundle.

Since the api implementation is independent of whether the setup was done via DD UI or admin, it would need to consider multiple matches possible.

Does that make sense?

Reply:
Ah. That does make sense. I was getting confused by the relationship between resource types and case types.

> Though we should add that as a validation on the model?

[We do raise `IntegrityError`](https://github.com/dimagi/commcare-hq/blob/6b4414b2cb1e298bfe08689681a05664a434013f/corehq/motech/fhir/models.py#L37) if someone tries to do something unsupported.

I've written a small test to show (mostly for myself) how that works: https://github.com/dimagi/commcare-hq/pull/29392

Reply:
Strange, I tried adding the same resource type for two different case types, and I could do that [here](https://staging.commcarehq.org/a/myfirsttestproject/data_dictionary/). That should not be possible then?

Reply:
Ah, I misunderstood. You can't have same case type mapped to two different resource types, but vice versa is okay?